File: javascript/src/mutual.js
/**
 * Mutual authentication challenge-response protocol for AgentPin.
 */

import { randomBytes } from 'crypto';
import { signData, verifySignature } from './crypto.js';

export const NONCE_EXPIRY_SECS = 60;

/**
 * Base64url encode bytes (no padding).
 * @param {Buffer} data
 * @returns {string}
 */
function base64urlEncode(data) {
    return data.toString('base64')
        .replace(/\+/g, '-')
        .replace(/\//g, '_')
        .replace(/=+$/, '');
}

/**
 * Create a challenge with a 128-bit random nonce.
 * @param {string|null} verifierCredential - Optional verifier credential JWT
 * @returns {object} Challenge object
 */
export function createChallenge(verifierCredential = null) {
    const nonceBytes = randomBytes(16); // 128 bits
    const nonce = base64urlEncode(nonceBytes);

    const challenge = {
        type: 'agentpin-challenge',
        nonce,
        timestamp: new Date().toISOString(),
    };

    if (verifierCredential) {
        challenge.verifier_credential = verifierCredential;
    }

    return challenge;
}

/**
 * Create a response by signing the challenge nonce.
 * @param {object} challenge
 * @param {string} privateKeyPem
 * @param {string} kid
 * @returns {object} Response object
 */
export function createResponse(challenge, privateKeyPem, kid) {
    const signature = signData(privateKeyPem, Buffer.from(challenge.nonce));

    return {
        type: 'agentpin-response',
        nonce: challenge.nonce,
        signature,
        kid,
    };
}

/**
 * Verify a challenge response: check signature and that nonce hasn't expired.
 * @param {object} response
 * @param {object} challenge
 * @param {string} publicKeyPem
 * @returns {boolean}
 * @throws {Error} if nonce has expired
 */
export function verifyResponse(response, challenge, publicKeyPem) {
    // Check nonce matches
    if (response.nonce !== challenge.nonce) {
        return false;
    }

    // Check timestamp hasn't expired
    const ts = new Date(challenge.timestamp);
    const elapsed = (Date.now() - ts.getTime()) / 1000;
    if (elapsed > NONCE_EXPIRY_SECS) {
        throw new Error(
            `Challenge nonce expired (${Math.floor(elapsed)} seconds old, max ${NONCE_EXPIRY_SECS})`
        );
    }

    // Verify signature over the nonce
    return verifySignature(publicKeyPem, Buffer.from(challenge.nonce), response.signature);
}

/**
 * Verify a challenge response with optional nonce deduplication.
 * @param {object} response
 * @param {object} challenge
 * @param {string} publicKeyPem
 * @param {import('./nonce.js').InMemoryNonceStore|null} nonceStore - Optional nonce store for replay prevention
 * @returns {boolean}
 * @throws {Error} if nonce has been replayed or expired
 */
export function verifyResponseWithNonceStore(response, challenge, publicKeyPem, nonceStore = null) {
    if (nonceStore) {
        if (!nonceStore.checkAndRecord(response.nonce, NONCE_EXPIRY_SECS * 1000)) {
            throw new Error(`Nonce '${response.nonce}' has already been used (replay attack)`);
        }
    }
    return verifyResponse(response, challenge, publicKeyPem);
}

File: python/agentpin/mutual.py
"""Mutual authentication challenge-response protocol for AgentPin."""

import base64
import os
from datetime import datetime, timezone
from typing import Optional

from .crypto import sign_data, verify_signature

NONCE_EXPIRY_SECS = 60


def _base64url_encode(data: bytes) -> str:
    """Base64url encode without padding."""
    return base64.urlsafe_b64encode(data).rstrip(b"=").decode("utf-8")


def create_challenge(verifier_credential: Optional[str] = None) -> dict:
    """Create a challenge with a 128-bit random nonce."""
    nonce_bytes = os.urandom(16)  # 128 bits
    nonce = _base64url_encode(nonce_bytes)

    challenge: dict = {
        "type": "agentpin-challenge",
        "nonce": nonce,
        "timestamp": datetime.now(timezone.utc).isoformat(),
    }

    if verifier_credential:
        challenge["verifier_credential"] = verifier_credential

    return challenge


def create_response(challenge: dict, private_key_pem: str, kid: str) -> dict:
    """Create a response by signing the challenge nonce."""
    signature = sign_data(private_key_pem, challenge["nonce"].encode("utf-8"))

    return {
        "type": "agentpin-response",
        "nonce": challenge["nonce"],
        "signature": signature,
        "kid": kid,
    }


def verify_response(response: dict, challenge: dict, public_key_pem: str) -> bool:
    """Verify a challenge response: check signature and that nonce hasn't expired.

    Raises:
        ValueError: if nonce has expired.
    """
    # Check nonce matches
    if response["nonce"] != challenge["nonce"]:
        return False

    # Check timestamp hasn't expired
    ts = datetime.fromisoformat(challenge["timestamp"])
    if ts.tzinfo is None:
        ts = ts.replace(tzinfo=timezone.utc)
    elapsed = (datetime.now(timezone.utc) - ts).total_seconds()
    if elapsed > NONCE_EXPIRY_SECS:
        raise ValueError(f"Challenge nonce expired ({int(elapsed)} seconds old, max {NONCE_EXPIRY_SECS})")

    # Verify signature over the nonce
    return verify_signature(public_key_pem, challenge["nonce"].encode("utf-8"), response["signature"])
